### ISSUE TITLE
Provided Pascal evaluation code get very diffferent results with Official Pascal Evaluation Method?

### ISSUE THREAD

**Issue by xuw080** [facebookresearch/maskrcnn-benchmark#258]
## ❓ Questions and Help
When I use provided voc evaluation code for evaluating results of Pascal Voc(trained with 07+12trainval and test on 07test), I use your codes will get 82.6 mAP(Faster-RCNN-FPN-ResNet101). However, if I use official evaluation codes provided in https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/datasets/voc_eval.py, I got different results, not sure whether you can check it and support using official evaluation codes? Thank you so much

**Comment by fmassa** [comment]
Hi,

One thing to verify is that there are two modes for evaluation: using [voc2007 metric or using voc2012 metric](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/data/datasets/evaluation/voc/voc_eval.py#L48).
From experience, both metrics give different results, as the 07 metric is much coarser when evaluating.

Can you see if this is the reason for the difference? If not, then I'd be very interested in get the differences removed.

**Comment by fmassa** [comment]
Hi,

Thanks for pointing this out.

I haven't myself implemented those evaluation metrics, it as a contributor that added this feature.

We should definitely find the reason why there is so much discrepancy, those functions where originally taken from py-faster-rcnn, so there shouldn't be much difference.

**Comment by fmassa** [comment]
cc @lufficc who originally sent the PR with the implementation

**Comment by xuw080** [comment]
Hi, I got different results. I train with 07trainval+12trainval and tested on 07 test. I trained with Faster-RCNN-R-50-C4 got 79.6 mAP(voc evaluation), 73.3 mAP(official evaluation) and 43.3 mAP, 72.7 mAP@0.5 and 46.6 mAP@0.75(coco evaluation). For Faster-RCNN-R-50-FPN, I got 81.34 mAP(voc evaluation), 74.3 mAP(official evaluation), and 45 mAP, 74.1 mAP@0.5 and 48.6 mAP@0.75(coco evaluation). For Faster-RCNN-R-101-FPN, I got 82.56 mAP(voc evaluation), 76.2 mAP(official evaluation), and 47.7 mAP, 75.5 mAP@0.5, 52.5 mAP@0.75. For Faster-RCNN-ResXt-101-FPN, I got 83.19 mAP(voc evaluation), 75.9 mAP(official evaluation), and 48.1 mAP, 75.1 mAP@0.5, 53.2 mAP@0.75. All codes are trained with 8 gpus, each holds two images. lr=0.02 and weight decay is 10k and 14.5k.(decay in 5 epochs and max is ~7 epochs, epochs number is the same as provided config file). I didn't change any codes.

**Comment by lufficc** [comment]
Do you use official code correctly？Could I have a look at your code to evaluate using official code? Then we may find the problem，because I got similar result when using official code.

**Comment by xuw080** [comment]
I will save all the detection results of each class to be in the format of:
image_id + " " + score + " " + xmin + " " + ymin + " " + xmax + " " + ymax + "\n". 
Then save all the same class predictions in the same txt file as 'ClassName'_test.txt. Then simply use official voc_eval.py to read all these files to get the official evaluation results. 

I am not sure whether you can send your evaluation code to me, I can test it with your implementation methods and provide the new testing results on all of my experiments. 

In current voc evaluation methods, Faster-RCNN-ResNet50's mAP reached 0.7947, which is much higher than any publication results. The highest results for Faster-RCNN-ResNet101 in published work is 79.8, ResNet50 should reach about 76~77 mAP. So, I don't think 79.47 is reasonable, also, I tested with COCO evaluation methods(using coco format provided in https://github.com/facebookresearch/multipathnet/blob/master/README.md), the mAP@0.5 for coco evaluation methods, i.e, 72.7 mAP@0.5, is also very different from the voc evaluation results. I am not sure whether I can have a look at your

**Comment by lufficc** [comment]
Replace this https://github.com/facebookresearch/maskrcnn-benchmark/blob/55d3ab44aa9ff5261afb8244ab5e265266d8436f/maskrcnn_benchmark/data/datasets/evaluation/voc/voc_eval.py#L12 with:
```python
def do_voc_evaluation(dataset, predictions, output_folder, logger):
    class_boxes = {dataset.map_class_id_to_class_name(i + 1): [] for i in range(20)}
    for image_id, prediction in tqdm(enumerate(predictions)):
        img_info = dataset.get_img_info(image_id)
        if len(prediction) == 0:
            continue
        image_width = img_info["width"]
        image_height = img_info["height"]
        prediction = prediction.resize((image_width, image_height))
        pred_bbox = prediction.bbox.numpy()
        pred_label = prediction.get_field("labels").numpy()
        pred_score = prediction.get_field("scores").numpy()

        for i, class_id in enumerate(pred_label):
            image_name = dataset.get_origin_id(image_id)
            box = pred_bbox[i]
            score = pred_score[i]
            class_name = dataset.map_class_id_to_class_name(class_id)
            class_boxes[class_name].append((image_name, box[0], box[1], box[2], box[3], score))
    aps = []
    tmp = os.path.join(output_folder, 'tmp')
    if not os.path.exists(tmp):
        os.makedirs(tmp)
    for key in dataset.CLASSES[1:]:
        filename = os.path.join(output_folder, '{}.txt'.format(key))
        if os.path.exists(filename):
            os.remove(filename)
        with open(filename, 'wt') as txt:
            boxes = class_boxes[key]
            for k in range(len(boxes)):
                box = boxes[k]
                txt.write('{:s} {:.3f} {:.1f} {:.1f} {:.1f} {:.1f}\n'.format(box[0], box[-1], box[1], box[2], box[3], box[4]))
        devkit_path = '/data7/lufficc/voc/VOCdevkit/VOC2007'
        annopath = os.path.join(devkit_path, 'Annotations', '{:s}.xml')
        imagesetfile = os.path.join(devkit_path, 'ImageSets', 'Main', 'test.txt')
        rec, prec, ap = voc_eval(filename, annopath, imagesetfile, key, tmp, ovthresh=0.5, use_07_metric=True)
        aps += [ap]
        print(('AP for {} = {:.4f}'.format(key, ap)))
    print(('Mean AP = {:.4f}'.format(np.mean(aps))))
    print('~~~~~~~~')
    print('Results:')
    for ap in aps:
        print(('{:.3f}'.format(ap)))
    print(('{:.3f}'.format(np.mean(aps))))
    print('~~~~~~~~')
    print('')
    print('--------------------------------------------------------------')
    print('Results computed with the **unofficial** Python eval code.')
    print('Results should be very close to the official MATLAB eval code.')
    print('Recompute with `./tools/reval.py --matlab ...` for your paper.')
    print('-- Thanks, The Management')
    print('--------------------------------------------------------------')
```

**Comment by xuw080** [comment]
Thanks, I will test it now.

**Comment by xuw080** [comment]
Hi, I compared official evaluation methods and your methods with two different datasets Pascal VOC0712 and KITTI, I will use voc style format for training and testing both of them. For pascal voc0712, the difference between your method and official evaluation method is in the range of 0.3~0.5 mAP. For KITTI, the difference between official evaluation method and your method will be over 10~20 mAP. In fact, originally, I thought maybe the evaluation method is different from the official one is because I got very different results on KITTI Official Benchmark even though I got almost same mAP with your voc evaluation methods as my previous results got from other implementation pytorch-faster-rcnn. I double checked your method and your provided official evaluation codes, I didn't find the reason, but the gap exists, which are smaller for Pascal and larger for KITTI. Following is my testing results using your original evaluation code and your official evaluation code:

Original:
2018-12-11 15:05:36,741 maskrcnn_benchmark.inference INFO: mAP: 0.5585
pedestrian      : 0.5833
cyclist         : 0.4626
car             : 0.8606
dontcare        : 0.3273

Official:
AP for pedestrian = 0.3188
AP for cyclist = 0.2444
AP for car = 0.4326
AP for dontcare = 0.1824
Mean AP = 0.2945

The official evaluation results matched the results I got in KITTI Official Benchmark.
I didn't find the reason, but I think maybe we should stick with official evaluation methods as what you provided in previous answers. Anyhow, thank you so much for supplying these codes and kind reply, It really helped me a lot.

**Comment by fmassa** [comment]
One question @xuw080 : are you using python 2?
It might be a matter of integer / float division differences between different python versions?

**Comment by xuw080** [comment]
I am using python3.7.

**Comment by fmassa** [comment]
@lufficc were you using python 2 when evaluating the code?
Also, could you try once more to see if you manage to reproduce the same results as before, or if you get results similar to @xuw080 ?
Thanks!

**Comment by lufficc** [comment]
I am using python3.6. I tried again and got very close result. And I use the same code in my another project which got very close result compare to the official code. I cannot fingure out why @xuw080 got so big difference on KITTI dataset.

**Comment by lufficc** [comment]
Maybe the two lines below cause the huge difference? Because if no object detected, will not do evaluating.
https://github.com/facebookresearch/maskrcnn-benchmark/blob/55d3ab44aa9ff5261afb8244ab5e265266d8436f/maskrcnn_benchmark/data/datasets/evaluation/voc/voc_eval.py#L19-L20
@fmassa @xuw080

**Comment by fmassa** [comment]
That might be the reason? I'd need to compare against the original implementation, but good spot!
Could you see if removing it bring the results closer?

**Comment by lufficc** [comment]
Now I am sure is the two line cause this.
After I change it to :
```python
if len(prediction) == 0 or image_id % 2 == 0:
    continue
```
The official code got 0.3840, ours got 0.7252. Without change to this, still close.
 So If It didn't detect any object on many images, Official code got low mAP, ours still got higher because not take in account them, this why @xuw080 got higher mAP on KITTI.
@fmassa

**Comment by fmassa** [comment]
We should remove the `if len(prediction) == 0` skipping, this is probably not right...

**Comment by lufficc** [comment]
Yee, We should remove the `if len(prediction) == 0 skipping`.

**Comment by fmassa** [comment]
does this break the rest of the evaluation pipeline?

**Comment by lufficc** [comment]
No, it doesn't.

**Comment by fmassa** [comment]
@drcege I was on holidays and didn't have the chance to look into it.
@lufficc do you think that the changes that we have mentioned could be the fix? If yes, could you send a PR with the fix?

Thanks!

**Comment by dby2017** [comment]
Now the evaluation pipeline is right?

**Comment by fmassa** [comment]
@dby2017 should hae been fixed in https://github.com/facebookresearch/maskrcnn-benchmark/pull/648
